### PR TITLE
Toyzero dataset simple

### PR DESCRIPTION
**Pull request by usert5432**
Simple dataset implementation that we discussed 2 weeks ago

### CODE REVIEW

**Review by brettviren — APPROVED** [review]

File: toytools/datasets/simple_toyzero.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        If specified, then crop images down to shape specified by `crop_shape`.
+        Default: None
+    expansion_factor : int
+        Number of cropped regions to select from a single image.
+        Default: 1
+    is_train : bool
+        `is_train` flag controls overall behavior of the
+        `SimpleToyzeroDataset`. If `is_train` is True, then the cropped regions
+        of toyzero images are selected at random. Otherwise, the cropped
+        regions are selected in a repdorucible manner using `seed` parameter.
```
Comment (pphuangyi) [inline toytools/datasets/simple_toyzero.py]: reproducible

File: toytools/datasets/simple_toyzero.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        self._prg.shuffle(indices)
+
+        val_size = self._val_size
+
+        if val_size <= 1:
+            val_size = int(len(indices) * val_size)
+        else:
+            val_size = val_size // self._efactor
+
+        train_size = len(indices) - val_size
```
Comment (pphuangyi) [inline toytools/datasets/simple_toyzero.py:103]: Should we assert that `val_size` is smaller than `len(indices)`?

File: toytools/datasets/simple_toyzero.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+    def _split_dataset(self):
+        """Split images into training/validation samples."""
+        indices = np.arange(len(self._images))
+        self._prg.shuffle(indices)
+
+        val_size = self._val_size
+
+        if val_size <= 1:
+            val_size = int(len(indices) * val_size)
```
Comment (pphuangyi) [inline toytools/datasets/simple_toyzero.py:99]: I might be misunderstanding something, but why, when the input `val_size` is a fraction, the integer `val_size` doesn't have to divide `self._efactor`, but when `val_size` is an integer, it does?

File: toytools/datasets/simple_toyzero.py (hunk 2 above)
Comment (pphuangyi) [inline toytools/datasets/simple_toyzero.py:103]: Would it be better if we ensure the resulting `val_size` is at least one? I might miss something, but what would happen if the `self._efactor` is bigger than the input `val_size`?

File: toytools/datasets/simple_toyzero.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            image_fake, self._prg, bkg_value, MIN_PIXEL_DIFF,
+            self._crop_shape, CROP_RETRIES
+        )
+
+        return (cropped_fake_image, crop_image(image_real, crop_region))
+
+    def __getitem__(self, index):
+
+        if self._is_train:
+            self._prg    = np.random.default_rng()
```
Comment (pphuangyi) [inline toytools/datasets/simple_toyzero.py]: So, when `self._is_train=True`, we don't really get an entry by index, but rather a totally random entry, right?

File: toytools/datasets/simple_toyzero.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+            sample_index = self._prg.choice(self._indices, size = 1)[0]
+        else:
+            self._prg    = np.random.default_rng(index)
```
Comment (pphuangyi) [inline toytools/datasets/simple_toyzero.py]: I don't quite understand why we use `index` as the seed to the `default_rng`.

Comment (pphuangyi) [reply toytools/datasets/simple_toyzero.py]: And we don't use the `self._prg` in any way, right?

File: toytools/datasets/simple_toyzero.py (hunk 1 above)
Comment (usert5432) [reply toytools/datasets/simple_toyzero.py]: Fixed. Thank you!

File: toytools/datasets/simple_toyzero.py (hunk 2 above)
Comment (usert5432) [reply toytools/datasets/simple_toyzero.py:103]: Yes, probably we should. As @ YHRen  suggested, I will rewrite train_test_split function and move it outside of datasets module. And when I do that I will handle the case of `val_size` >= `len(indices)`. Here is the issue to track the progress https://github.com/LS4GAN/toytools/issues/6#issue-948658169.

File: toytools/datasets/simple_toyzero.py (hunk 3 above)
Comment (usert5432) [reply toytools/datasets/simple_toyzero.py:99]: Yes, sorry, this is counterintuitive behavior, and probably I should remove it. The motivation was, that `efactor` indicates how many cropped regions to sample from a single image. So, if we want our validation dataset to have 1000 cropped regions, then first we need to have `1000 / efactor` full images. And then sample `efactor` of cropped regions from each full image.

On the other hand, if we want our validation dataset to be 20% of the entire dataset, then we can simply select 20% of all full images and sample `efactor` of cropped regions from each.

File: toytools/datasets/simple_toyzero.py (hunk 2 above)
Comment (usert5432) [reply toytools/datasets/simple_toyzero.py:103]: Yep. I will put checks on `val_size` after refactoring https://github.com/LS4GAN/toytools/issues/6. I will probably handle `efactor` differently as well.

File: toytools/datasets/simple_toyzero.py (hunk 4 above)
Comment (usert5432) [reply toytools/datasets/simple_toyzero.py]: Yes, but I am not sure if that is a good decision. So, I am open to suggestions.

File: toytools/datasets/simple_toyzero.py (hunk 5 above)
Comment (usert5432) [reply toytools/datasets/simple_toyzero.py]: > I don't quite understand why we use `index` as the seed to the `default_rng`.

I want validation sample to be reproducible. So, I seed prg by an index of a cropped_region.


> And we don't use the `self._prg` in any way, right?

This prg is then used to crop images in `_get_rand_cropped_area`. But I agree, that it is a bit obscure. I will rewrite it to make explicit.

Comment (usert5432) [reply toytools/datasets/simple_toyzero.py]: I have made `prg` handling more explicit in ea8407906f91773be06170c17d507fe288059dc0.

File: toytools/datasets/simple_toyzero.py (hunk 3 above)
Comment (pphuangyi) [reply toytools/datasets/simple_toyzero.py:99]: Hi Dmitrii, thank you for the explanation! I thought this must be the case when I read it yesterday, but it is, as you said, a little counterintuitive because `val_size` seems to mean two slightly different things here. I think maybe the explanation your give here is sufficient and if we want to keep the code.

File: toytools/datasets/simple_toyzero.py (hunk 4 above)
Comment (pphuangyi) [reply toytools/datasets/simple_toyzero.py]: Since we also shuffle the indices, I am wondering another random access to entry is necessary or not.

Comment (usert5432) [reply toytools/datasets/simple_toyzero.py]: Good point. Ok, I think I will merge this branch to speed up the process, but I will open an issue to refactor/rewrite the SimpleDataset
